src/dbnode/client/replicated_session.go

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/storage/block"
+	"github.com/m3db/m3/src/dbnode/storage/bootstrap/result"
+	"github.com/m3db/m3/src/dbnode/storage/index"
+	"github.com/m3db/m3/src/dbnode/storage/namespace"
+	"github.com/m3db/m3/src/dbnode/topology"
+	"github.com/m3db/m3x/ident"
+	m3sync "github.com/m3db/m3x/sync"
+	xtime "github.com/m3db/m3x/time"
+)
+
+const (
+	maxReplicationConcurrency = 128
+)
+
+// replicatedSession is an implementation of clientSession which replicates
+// session read/writes to a set of clusters asynchronously.
+type replicatedSession struct {
+	session       AdminSession
+	asyncSessions []AdminSession
+	workerPool    m3sync.WorkerPool
+}
+
+// Ensure replicatedSession implements the clientSession interface.
+var _ clientSession = (*replicatedSession)(nil)
+
+func newReplicatedSession(opts MultiClusterOptions) (clientSession, error) {
+	workerPool := m3sync.NewWorkerPool(maxReplicationConcurrency)
```

Comment:
Blocking until a goroutine is free isn't desirable behavior in our case right?

>      // Go assign the Work to be executed by a Goroutine. Whether or not
> 	// it waits for an existing Goroutine to become available or not
> 	// is determined by the GrowOnDemand() option. If GrowOnDemand is not
> 	// set then the call to Go() will block until a goroutine is available.
> 	// If GrowOnDemand() is set then it will expand the pool of goroutines to
> 	// accommodate the work. The newly allocated goroutine will temporarily
> 	// participate in the pool in an effort to amortize its allocation cost, but
> 	// will eventually be killed. This allows the pool to dynamically respond to
> 	// workloads without causing excessive memory pressure. The pool will grow in
> 	// size when the workload exceeds its capacity and shrink back down to its
> 	// original size if/when the burst subsides

Reply:
@simonrobb Ah I just realized that the `GoIfAvailable` method is not present on the `PooledWorkerPool` implementation. Can you just leave a TODO to replace this with a pooled work pool once its updated to have a `GoIfAvailable` method? Should be easy to add but don't need to do it in this P.R

Reply:
Cool sounds good.